web3/middleware/signing.py

```diff
@@ -0,0 +1,62 @@
+from eth_keys.datatypes import PrivateKey
+from eth_keys.exceptions import ValidationError
+
+from web3.utils.datastructures import HexBytes
+
+import logging
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def construct_transaction_signing_middleware(private_key):
+    """Intercept a transaction in flight and sign it.
+
+    Args:
+        private_key (TYPE): Private key used to sign the transaction.
+                            Accepts the following formats:
+                            - raw bytes
+                            - hex encoded string
+                            - `PrivateKey` object from `eth-keys`
+                            - `LocalAccount` object from `web3.eth.account`
+
+    Returns:
+        TYPE: Description
+    """
+
+    # Check if the private_key is either
+    # - eth_key.keys.PrivateKey
+    # - web3.eth.account.create().privateKey
+    # - bytes
+    valid_key_types = (PrivateKey, HexBytes, bytes)
```

Comment:
`HexBytes` inherits from `bytes`, so an explicit `HexBytes` check is not necessary here. (Although a check against `LocalAccount` would be nice)

Reply:
roger that.  Need to add `LocalAccount` in there.

Reply:
OK, I remember the issue I was having.  In the transaction, just the regular address is passed, right?  The `acct.address` returns a "HexBytes" object, which looks similar to a plain string.  It seems to lose its class information.

```
In [34]: acct = w3.eth.account.create()
In [35]: transaction = {'from': acct.address, 'to': 'testrpc'}
In [37]: isinstance(acct.privateKey, web3.utils.signing.LocalAccount)
Out[37]: False
In [38]: isinstance(acct.address, web3.utils.signing.LocalAccount)
Out[38]: False
In [39]: isinstance(acct, web3.utils.signing.LocalAccount)
Out[39]: True
```

Any suggestions for an elegant way to check that the `acct.address` was in fact generated by `web3.eth.account.create()`?

Reply:
Github marked this as outdated -- not true.  Mentioning @carver so it doesn't get lost in the noise :)

Reply:
> The acct.address returns a "HexBytes" object

That would definitely be a bug, `acct.address` should be a checksummed hex string with the account address. That's what I get at the command line in a quick test right now. Can you show how to reproduce in an issue if you are seeing `HexBytes` in `acct.address`?

> Any suggestions for an elegant way to check that the acct.address was in fact generated by web3.eth.account.create()?

`transaction['from'] == acct.address` will tell you if the `acct` has the private key for the sender of `transaction`. But I feel like I might be missing something about what you're asking, so can you rephrase if I misunderstood?

Reply:
Or `eth_utils.is_same_address` as Piper mentioned later. (Although I think there are no code paths where addresses without a checksum here is possible, it can't hurt)

Reply:
@carver I misspoke.  Meant to say that the `acct.privateKey` returns the HexBytes.

```
In [7]: acct.privateKey
Out[7]: HexBytes('0xab6f1f3e4327d2cf3aac02abd898ded798a57f5deab603ac16597686eddaeeda')
In [8]: type(acct.privateKey)
Out[8]: web3.utils.datastructures.HexBytes
```

So what I was asking was the best way to check that the privateKey was generated from the `web3.eth.account.create()` method.

Reply:
Maybe the simplest interface is to do `Account.privateKeyToAccount(key).address`